### PR TITLE
feat: Implement user confirmation mode, request confirmation when running bash/python code in this mode

### PR DESCRIPTION
**Pull request by adrgs**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**
Implements #2308 enhancement

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**
- Adds a new boolean setting called CONFIRMATION_MODE, which is false by default
- Adds a new toggle option in SettingsForm
- Implements 3 new states for the agents: `AWAITING_USER_CONFIRMATION`, `ACTION_CONFIRMED`, `ACTION_REJECTED`
- In confirmation mode, the bash/python actions will have confirm/reject buttons directly in the chat log
- Updated tests

**Other references**

<details><summary>Screenshots:</summary>
Screenshot 2024-07-03 at 17 02 46

Screenshot 2024-07-03 at 18 28 17

Screenshot 2024-07-03 at 17 02 59

Screenshot 2024-07-03 at 17 04 17

Screenshot 2024-07-03 at 18 12 00

Screenshot 2024-07-03 at 18 12 08
</details>

### CODE REVIEW

**Review by xingyaoww — COMMENTED** [review]
Backend part LGTM 👍 ! Can anyone from the the frontend side take a look?

**Comment by iFurySt** [comment]
LGTM! @adrgs , could you please fix the lint and the failed unit tests?

File: opendevin/controller/agent_controller.py [hunk 1]
```diff
@@ -176,9 +182,19 @@ async def on_event(self, event: Event):
...
+                and self._pending_action.is_confirmed
+                == ActionConfirmationStatus.AWAITING_CONFIRMATION
+            ):
+                return
             if self._pending_action and self._pending_action.id == event.cause:
                 await self.add_history(self._pending_action, event)
                 self._pending_action = None
+                if self.state.agent_state == AgentState.ACTION_CONFIRMED:
+                    await self.set_agent_state_to(AgentState.RUNNING)
+                if self.state.agent_state == AgentState.ACTION_REJECTED:
```
Comment (enyst) [inline opendevin/controller/agent_controller.py:196]: Thanks for this, really cool addition. ❤️ 

Just a little detail here: we have in the codebase another REJECT action, with the meaning that a delegate agent rejects a task it's been given. That one involves some 40 more occurrences of "Reject"-action, enum, etc, related to that meaning, all throughout the code.

I wonder if we can avoid the potential confusion about which is which. If we consider them different enough, then maybe we can name one DENY instead?

Or if they have enough in common, like, this is a user rejection, the other is an agent rejection, maybe we can name with prefixes like 'user' and 'agent'. It seems to me the first is more likely, but I'm not fully sure. Cc: @li-boxuan

Comment (li-boxuan) [reply opendevin/controller/agent_controller.py:196]: I like the prefix idea. How about renaming this to `AgentState.USER_REJECTED` and `AgentState.USER_CONFIRMED`?

File: frontend/src/services/settings.ts [hunk 2]
```diff
@@ -5,16 +5,22 @@ export type Settings = {
...
 export const DEFAULT_SETTINGS: Settings = {
   LLM_MODEL: "gpt-4o",
   AGENT: "CodeActAgent",
   LANGUAGE: "en",
   LLM_API_KEY: "",
+  CONFIRMATION_MODE: false,
 };
 
-const validKeys = Object.keys(DEFAULT_SETTINGS) as (keyof Settings)[];
+const validKeys = Object.keys(DEFAULT_SETTINGS) as (keyof SettingsInput)[];
```
Comment (amanape) [inline frontend/src/services/settings.ts:23]: This is an incorrect change, types such as `string | boolean` do not have keys, `validKeys` end up with type `"valueOf"[]` (inferred from the common methods between `String.prototype` and `Boolean.prototype`)

```suggestion
const validKeys = Object.keys(DEFAULT_SETTINGS) as (keyof Settings)[];
```
The code now correctly infers the types of `validKeys` to be `(keyof Settings)[]`

File: frontend/src/services/settings.ts [hunk 3]
```diff
@@ -66,10 +74,10 @@ export const getSettings = (): Settings => {
  */
 export const saveSettings = (settings: Partial<Settings>) => {
   Object.keys(settings).forEach((key) => {
-    const isValid = validKeys.includes(key as keyof Settings);
+    const isValid = validKeys.includes(key as keyof SettingsInput);
```
Comment (amanape) [inline frontend/src/services/settings.ts:77]:
```suggestion
    const isValid = validKeys.includes(key as keyof Settings);
```

File: frontend/src/services/settings.ts [hunk 4]
```diff
@@ -91,11 +99,12 @@ export const getSettingsDifference = (settings: Partial<Settings>) => {
   const updatedSettings: Partial<Settings> = {};
 
   Object.keys(settings).forEach((key) => {
+    const typedKey = key as keyof SettingsInput;
```
Comment (amanape) [inline frontend/src/services/settings.ts:102]:
```suggestion
    const typedKey = key as keyof Settings;
```

File: frontend/src/services/settings.ts [hunk 5]
```diff
@@ -91,11 +99,12 @@ export const getSettingsDifference = (settings: Partial<Settings>) => {
...
     if (
-      validKeys.includes(key as keyof Settings) &&
-      settings[key as keyof Settings] !== currentSettings[key as keyof Settings]
+      validKeys.includes(typedKey) &&
+      settings[typedKey] !== currentSettings[typedKey]
     ) {
-      updatedSettings[key as keyof Settings] = settings[key as keyof Settings];
+      updatedSettings[typedKey] = settings[typedKey];
```
Comment (amanape) [inline frontend/src/services/settings.ts:107]:
```suggestion
      (updatedSettings[typedKey] as SettingsInput) = settings[typedKey] as SettingsInput;
```

**Comment by adrgs** [comment]
Thanks everyone for the great feedback! I've added a new commit with the changes.

Let me know if there is anything else to be done. Thank you

**Review by enyst — APPROVED** [review]

**Comment by enyst** [comment]
I think CmdKillAction no longer exists, merge issue?

File: opendevin/controller/agent_controller.py (hunk 1 above)
Comment (adrgs) [reply opendevin/controller/agent_controller.py]: Added changes to [2f510dc](https://github.com/OpenDevin/OpenDevin/pull/2774/commits/2f510dc71094d2bd89bfae6d3391467afc63ace1)

File: frontend/src/components/chat/Chat.tsx [hunk 6]
```diff
@@ -1,15 +1,26 @@
...
 import ChatMessage from "./ChatMessage";
+import { RootState } from "#/store";
+import AgentState from "#/types/AgentState";
 
 interface ChatProps {
   messages: Message[];
 }
 
 function Chat({ messages }: ChatProps) {
+  const { curAgentState } = useSelector((state: RootState) => state.agent);
```
[the remark this answers is not in the document]
Comment (adrgs) [reply frontend/src/components/chat/Chat.tsx]: Added changes to [2f510dc](https://github.com/OpenDevin/OpenDevin/pull/2774/commits/2f510dc71094d2bd89bfae6d3391467afc63ace1)

**Review by amanape — CHANGES_REQUESTED** [review]
Hey! Sorry for the delay. It seems your `package-lock.json` and `package.json` are not in sync. Try removing the `package-lock.json` and running `npm install` again.

**Comment by adrgs** [comment]
Added, sorry for that.

**Comment by amanape** [comment]
It looks that one of the Python tests are failing. I'm not sure if it is due to flakiness or due to the changes you have introduced.

@enyst Could you confirm if it is due to the change they made to `agent.py`?

**Comment by yufansong** [comment]
I guess it should be flakiness. Rerun the test to see the result.

**Comment by adrgs** [comment]
Added missing is_confirmed in test outputs

**Review by amanape — APPROVED** [review]
Thanks a lot for this cool feature! Awesome stuff

**Pull request merged**